### ISSUE TITLE
Thanks!

### ISSUE THREAD

**Issue by coderschoolreview** [tuanhung303/CoderSchool#1]
## Goal of this Assignment
The goal of this assignment was to introduce you to 2 main concepts in Machine Learning: 
- Data Visualization
- Exploratory Data Analysis. 

You learn how to query and clean data using pandas library in Python, make some plots which help to understand more about data with Seaborn library.

## Things you did well:
- You accomplished every questions. This is a great start !
- Your solution is very comprehensive and shows that you know how to utilize Pandas very well.
- You also perform the bonus dataset. Awesome !!!
- From the bonus dataset, you draw some conclusions from it. And the great thing is being curious about the dataset and make conclusions to confirm the hypothesis. Keep up !

## To sum up:
- Overall, excellent work! You are demonstrating that you are understanding the material and doing a great job of applying it. Keep it up!

**Comment by coderschoolreview** [comment]
The goal of this assignment is for you to learn to build and evaluate a model using scikit-learn library in Python. You learn how to do Data pre-processing, splitting it into training and test sets, training your model using training sets and evaluating its performance on test sets.

Things you did well: 
- Discovering the usage of many new functions!
- Understanding how to use data to train a model using library in Python. This is a very important step!

Some minor tips:
- You should take more features into your X (before splitting it), instead of just 1 feature “Time On App”.
- You don’t actually need to sort the X_train, X_test, y_train and y_test, unless you are trying to plot them. 
- In the “Residual” section, you count the different between your y_test and predictions, and then put it into graph to see whether it is a normal distribution. So: residuals = y_test-predictions, and then plot it using: sns.distplot(residuals,bins=10)
- The meaning of those coefficients is that when 1 feature increase (decrease) by 1, it changes your outcome (in this case Yearly Amount Spent) by 1*value_of_coefficient.

**Comment by coderschoolreview** [comment]
Hi Hung Tuan,

Excellent work on your Assignment 3. You seem confident with performing tasks that include sentiment analysis.

Keep it up!

**Comment by coderschoolreview** [comment]
# hyl0428: Assignment 5

The goal of this assignment was to introduce you following concepts in Machine Learning:

- Data quality check: missing value, anomaly, ...
- Exploratory Data Analysis: ratio of class, relationship of feature with another and relationship of feature with target, and data visualization.
- Preprocessing data: handling imbalance dataset, dealing with categorical variable.
- Modeling & evaluation: building classifier and evaluate, Hyper Parameter Tuning.

## Things you did well
* Aweomse, you knew magic function `!pip install missingno`, more on magic command [here](https://jakevdp.github.io/PythonDataScienceHandbook/01.03-magic-commands.html)
* Checking data quality: check for missing value and using `missingno` to visualize
* EDA: Ratio of classes, finding out that data is imbalance, visualizing relationship between features 
* Preprocessing: under sampling / over sampling, train test split
* Modeling & evaluate: train and evaluate multiple model by comparing classification report of models against each other

## Things to work on
* Preprocessing categorical features. Try `pandas.get_dummies`
* GridSearchCV and RandomSearchCV: common practice is using RandomSearchCV to narrow down search range then fine tuning with GridSearchCV, can read [this article](https://towardsdatascience.com/hyperparameter-tuning-the-random-forest-in-python-using-scikit-learn-28d2aa77dd74)

## Minor tips

* To know "**which is the best combination of parameter**":

  https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GridSearchCV.html

  ```python
  # Using gridsearchcv, random forest model and this param grid to find the best combination of parameters 
  # Hint: example
  # https://stackoverflow.com/questions/30102973/how-to-get-best-estimator-on-gridsearchcv-random-forest-classifier-scikit
  
  param_grid = {
      'bootstrap': [True],
      'max_depth': [80, 90, 100, 110],
      'max_features': [2, 3],
      'min_samples_leaf': [3, 4, 5],
      'min_samples_split': [8, 10, 12],
      'n_estimators': [100, 200, 300, 1000]
  }
  
  
  from sklearn.model_selection import GridSearchCV
  gcv = GridSearchCV(RandomForestClassifier(),param_grid=param_grid)
  gcv.fit(X,y)
  gcv.best_params_
  ```

* Installing packages `missingno ` (or any arbitrary package) on Win10:

  * Go to `Anaconda Prompt`
  * Type `conda install -c conda-forge missingno`

* Filtering series: 

  * Your code:`train_copy.isnull()`

  * Could write like this to get column with null value only:

    ```python
    ncols = train_copy.isnull().sum
    ncols[ncols!=0]
    ```

* Check if whole data frame have any null value:

  ​	`train.isnull().any().any()`

* For evaluation function, should `print` instead of `return` so when you loop through list of model and evaluate them, the result for each iteration printed to output.

**Comment by coderschoolreview** [comment]
<div style="background-color:pink;padding:2%"> 
<b>Please ignore previous comment, commented on wrong profile, terribly sorry</b>
</div>
# Hung Tuan: Assignment 5

The goal of this assignment was to introduce you following concepts in Machine Learning:

- Data quality check: missing value, anomaly, ...
- Exploratory Data Analysis: ratio of class, relationship of feature with another and relationship of feature with target, and data visualization.
- Preprocessing data: handling imbalance dataset, dealing with categorical variable.
- Modeling & evaluation: building classifier and evaluate, Hyper Parameter Tuning.

## Things you did well
* Checking data quality: check for missing value and using `missingno` to visualize
* EDA: Ratio of classes, finding out that data is imbalance, visualizing relationship between features 
* Preprocessing: under sampling / over sampling, train test split
* Modeling & evaluate: train and evaluate multiple model by comparing classification report of models against each other

## Things to work on
* Preprocessing categorical features. Try `pandas.get_dummies`
* GridSearchCV and RandomSearchCV: common practice is using RandomSearchCV to narrow down search range then fine tuning with GridSearchCV, can read [this article](https://towardsdatascience.com/hyperparameter-tuning-the-random-forest-in-python-using-scikit-learn-28d2aa77dd74)

## Minor tips


* Installing packages `missingno ` (or any arbitrary package) on Win10:

  * Go to `Anaconda Prompt`
  * Type `conda install -c conda-forge missingno`

Or, in the notebook type: `!pip install missingno`

* Filtering series: 

  * Your code:`train_copy.isnull()`

  * Could write like this to get column with null value only:

    ```python
    ncols = train_copy.isnull().sum
    ncols[ncols!=0]
    ```

* Check if whole data frame have any null value:

  ​	`train.isnull().any().any()`

* For evaluation function, should `print` instead of `return` so when you loop through list of model and evaluate them, the result for each iteration printed to output.

**Comment by coderschoolreview** [comment]
---
# Assignment 6

# Goal of this Assignment

The goal of this assignment was to introduce you to following concepts:
* Unsupervised Learning (KMeans, Hierachical Clustering)
* PCA

You learn how to use PCA for dimension reduction, KMeans, and Hierarchical Clustering. Also you learn to visualize the result of both tenichque.

# Things you did well:

* You accomplished every questions. This is a great start !
* Your solution is very comprehensive.

# To sum up:

Overall, excellent work! You are demonstrating that you are understanding the material and doing a great job of applying it. Keep it up!

# Few minor tips

> ![](http://i.imgur.com/TU7gOb6.png)
For this question, there's a shorter version. (What if there's a hundred features?)
```python
total_purchases = data.sum(axis=1)
purchase_percent = data.div(total_purchases, axis=0) * 100
```

> ![](http://i.imgur.com/qD4XUnJ.png)
For this question. you're plotting 2 first features of original dataset, while the question requires to plot on 2 first component of PCA. Try this
```python
plt.figure(figsize=(10, 7))
plt.scatter(pca_data['PC1'], pca_data['PC2'], c=cluster.labels_, cmap='rainbow');
```
